libcudacxx/test/libcudacxx/std/ranges/range.adaptors/range.all/range.all/all.pass.cpp

```diff
@@ -0,0 +1,290 @@
+  __host__ __device__ constexpr CopyableView& operator=(CopyableView const& other) noexcept(IsNoexcept)
+  {
+    start_ = other.start_;
+    return *this;
+  }
+  __host__ __device__ constexpr explicit CopyableView(int start) noexcept
+      : start_(start)
+  {}
+  __host__ __device__ constexpr int* begin() const
+  {
+    return globalBuff + start_;
+  }
+  __host__ __device__ constexpr int* end() const
+  {
+    return globalBuff + 8;
+  }
+};
+static_assert(cuda::std::ranges::view<CopyableView<true>>);
+static_assert(cuda::std::ranges::view<CopyableView<false>>);
+
+struct MoveOnlyView : cuda::std::ranges::view_base
+{
+  MoveOnlyView()                               = default;
+  MoveOnlyView(const MoveOnlyView&)            = delete;
+  MoveOnlyView& operator=(const MoveOnlyView&) = delete;
+  MoveOnlyView(MoveOnlyView&&)                 = default;
+  MoveOnlyView& operator=(MoveOnlyView&&)      = default;
+
+  __host__ __device__ int* begin() const;
+  __host__ __device__ int* end() const;
+};
+
+struct Range
+{
+  int start_;
+  __host__ __device__ constexpr explicit Range(int start) noexcept
+      : start_(start)
+  {}
+  __host__ __device__ constexpr int* begin() const
+  {
+    return globalBuff + start_;
+  }
+  __host__ __device__ constexpr int* end() const
+  {
+    return globalBuff + 8;
+  }
+};
+
+struct BorrowableRange
+{
+  int start_;
+  __host__ __device__ constexpr explicit BorrowableRange(int start) noexcept
+      : start_(start)
+  {}
+  __host__ __device__ constexpr int* begin() const
+  {
+    return globalBuff + start_;
+  }
+  __host__ __device__ constexpr int* end() const
+  {
+    return globalBuff + 8;
+  }
+};
+template <>
+inline constexpr bool cuda::std::ranges::enable_borrowed_range<BorrowableRange> = true;
+
+struct RandomAccessRange
+{
+  __host__ __device__ constexpr auto begin()
+  {
+    return random_access_iterator<int*>(globalBuff);
+  }
+  __host__ __device__ constexpr auto end()
+  {
+    return sized_sentinel(random_access_iterator<int*>(globalBuff + 8));
+  }
+};
+template <>
+inline constexpr bool cuda::std::ranges::enable_borrowed_range<RandomAccessRange> = true;
+
+#if TEST_STD_VER >= 2020
+template <class View, class T>
+concept CanBePiped = requires(View&& view, T&& t) {
+  { cuda::std::forward<View>(view) | cuda::std::forward<T>(t) };
+};
+#else // ^^^ C++20 ^^^ / vvv C++17 vvv
+template <class View, class T, class = void>
+inline constexpr bool CanBePiped = false;
+
+template <class View, class T>
+inline constexpr bool
+  CanBePiped<View, T, cuda::std::void_t<decltype(cuda::std::declval<View>() | cuda::std::declval<T>())>> = true;
+#endif // TEST_STD_VER <= 2017
+
+__host__ __device__ constexpr bool test()
+{
+  {
+    static_assert(cuda::std::is_same_v<decltype(cuda::std::views::all(View<true>())), View<true>>);
+    static_assert(noexcept(cuda::std::views::all(View<true>())));
+// old GCC seems to fall over the noexcept clauses here
+#if !TEST_COMPILER(GCC, <, 9) && !TEST_COMPILER(MSVC)
```

Comment:
Does MSVC fail as well?

Reply:
I checked 2022 and it passes will try another CI run with this removed

Reply:
rechecked and VS2019 is affected, so I restricted it to that